core/src/test/scala/com/softwaremill/sttp/testing/HttpTest.scala

```diff
@@ -0,0 +1,526 @@
+      val resp = r2.send().force()
+      resp.code should be(200)
+      resp.unsafeBody should be(r4response)
+    }
+
+    "redirect twice when redirects should be followed" in {
+      val resp = r1.send().force()
+      resp.code should be(200)
+      resp.unsafeBody should be(r4response)
+    }
+
+    "redirect when redirects should be followed, and the response is parsed" in {
+      val resp = r2.response(asString.map(_.toInt)).send().force()
+      resp.code should be(200)
+      resp.unsafeBody should be(r4response.toInt)
+    }
+
+    "keep a single history entry of redirect responses" in {
+      val resp = r3.send().force()
+      resp.code should be(200)
+      resp.unsafeBody should be(r4response)
+      resp.history should have size (1)
+      resp.history(0).code should be(302)
+    }
+
+    "keep whole history of redirect responses" in {
+      val resp = r1.send().force()
+      resp.code should be(200)
+      resp.unsafeBody should be(r4response)
+      resp.history should have size (3)
+      resp.history(0).code should be(307)
+      resp.history(1).code should be(308)
+      resp.history(2).code should be(302)
+    }
+
+    "break redirect loops" in {
+      val resp = loop.send().force()
+      resp.code should be(0)
+      resp.history should have size (FollowRedirectsBackend.MaxRedirects)
+    }
+
+    "break redirect loops after user-specified count" in {
+      val maxRedirects = 10
+      val resp = loop.maxRedirects(maxRedirects).send().force()
+      resp.code should be(0)
+      resp.history should have size (maxRedirects)
+    }
+
+    "not redirect when maxRedirects is less than or equal to 0" in {
+      val resp = loop.maxRedirects(-1).send().force()
+      resp.code should be(302)
+      resp.body should be('left)
+      resp.history should be('empty)
+    }
+  }
+
+  "timeout" - {
+    "fail if read timeout is not big enough" in {
+      val request = sttp
+        .get(uri"$endpoint/timeout")
+        .readTimeout(200.milliseconds)
+        .response(asString)
+
+      intercept[Throwable] {
+        request.send().force()
+      }
+    }
+
+    "not fail if read timeout is big enough" in {
+      val request = sttp
+        .get(uri"$endpoint/timeout")
+        .readTimeout(5.seconds)
+        .response(asString)
+
+      request.send().force().unsafeBody should be("Done")
+    }
+  }
+
+  "empty response" - {
+    val postEmptyResponse = sttp
+      .post(uri"$endpoint/empty_unauthorized_response")
+      .body("{}")
+      .contentType("application/json")
+
+    "parse an empty error response as empty string" in {
+      val response = postEmptyResponse.send().force()
+      response.body should be(Left(""))
+    }
+  }
+
+  "encoding" - {
+    "read response body encoded using ISO-8859-2, as specified in the header, overriding the default" in {
+      val request = sttp.get(uri"$endpoint/respond_with_iso_8859_2")
+
+      request.send().force().unsafeBody should be(textWithSpecialCharacters)
+    }
+  }
```

Comment:
I have manually checked the output of a test run against the previous `HttpTest` file, all tests are accounted for.

Reply:
👍